### ISSUE TITLE
Using an Upgrader with libp2p-webrtc-star

### ISSUE THREAD

**Issue by daviddahl** [libp2p/js-libp2p#650]
A per the discussion here: https://github.com/libp2p/js-libp2p-webrtc-star/issues/218#issuecomment-635455265

@vasco-santos asked me to file this issue to begin a discussion of how the `Upgrader` and `wrtc` should be / can be configured in the latest `libp2p` while running an application in node,js.

- **0.27.7**:
- **Linux 5.4.0-29-generic #33-Ubuntu SMP Wed Apr 29 14:32:27 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux**:
- **Transports**:

#### Description:

Adding a `libp2p-webrtc-star` instance into a Transports module configuration requires an `Upgrader` (instance? class?) - my clumsy attempt in this fork of IPFS examples: https://github.com/daviddahl/js-ipfs/pull/1 best explains what I am attempting to do with `libp2p` on node.js 

#### Steps to reproduce the error:

Clone the branch here: https://github.com/daviddahl/js-ipfs/pull/1 
`cd js-ipfs/examples/custom-libp2p && npm install && node index.js`

You should see:
```
> node index.js 
TypeError: Cannot read property 'Symbol(Symbol.toStringTag)' of undefined
    at /home/ddahl/projects/test/js-ipfs/node_modules/libp2p/src/index.js:118:38
    at Array.forEach (<anonymous>)
    at new Libp2p (/home/ddahl/projects/test/js-ipfs/node_modules/libp2p/src/index.js:117:29)
    at /home/ddahl/projects/test/js-ipfs/examples/custom-libp2p/index.js:61:17
    at processTicksAndRejections (internal/process/task_queues.js:89:5)
```

If you comment out this line: https://github.com/daviddahl/js-ipfs/pull/1/files#diff-3160a5b56836421c3b48dcebbf951bd5R67 for `libp2p` to run but not consume the `webrtc-star` instance.

I just noticed an `rc` release for `0.28`, does it impact this functionality?

**Comment by jacobheun** [comment]
Are you actually trying to customize the upgrader, or just pass wrtc to the webrtc-star constructor? 

We don't support customizing the upgrader, but it looks like you just need to customize webrtc-star.

You can pass in options to the transport via the libp2p configuration, there is an example in the config readme for webrtc-star, https://github.com/libp2p/js-libp2p/blob/v0.27.8/doc/CONFIGURATION.md#customizing-transports. This will automatically handle passing the existing upgrader into the transport.


> I just noticed an rc release for 0.28, does it impact this functionality?

It won't. It does deprecate PeerInfo, so the constructor will [slightly change](https://github.com/libp2p/js-libp2p/blob/v0.28.0-rc.0/doc/migrations/v0.27-v.28.md#using-an-existing-peerid), but this doesn't effect the functionality for forwarding transport config options.

**Comment by vasco-santos** [comment]
I think that we should improve the configurability of it per transport (maybe part of #576) and we should support a way of extending the upgrader current behaviour.

In this case, we want to add config options on it. We should just get a easy way of doing it for now, and then use #576 to provide full support

**Comment by daviddahl** [comment]
> Are you actually trying to customize the upgrader, or just pass wrtc to the webrtc-star constructor?

Thank you for this information!
I was not trying to customize it - I was just trying to configure webrtc-star, as the exception being thrown mentioned passing in (or I assumed that) an Upgrader object. 
 
> You can pass in options to the transport via the libp2p configuration, there is an example in the config readme for webrtc-star, https://github.com/libp2p/js-libp2p/blob/v0.27.8/doc/CONFIGURATION.md#customizing-transports. This will automatically handle passing the existing upgrader into the transport.

I was on 0.27.7 as well, maybe that was part of it?

> It won't. It does deprecate PeerInfo, so the constructor will [slightly change](https://github.com/libp2p/js-libp2p/blob/v0.28.0-rc.0/doc/migrations/v0.27-v.28.md#using-an-existing-peerid), but this doesn't effect the functionality for forwarding transport config options.

^^ good to know!

I'll try again following this documentation.

**Comment by daviddahl** [comment]
So after another attempt, it seems like this documentation: https://github.com/libp2p/js-libp2p-webrtc-star#using-this-module-in-nodejs-read-not-in-the-browser

conflicts with this https://github.com/libp2p/js-libp2p/blob/v0.27.8/doc/CONFIGURATION.md#customizing-transports

Also, does one need to run the webrtc-star server via a libp2p node - or along side a libp2p node? - which seems to be referred to as a `bootstrap` node? see: https://github.com/shresthagrawal/jslibp2p-chat-tutorial/blob/master/assets/5/5.2-finished-code.js#L45 and https://github.com/shresthagrawal/jslibp2p-chat-tutorial/blob/master/assets/5/5.2-finished-code.js#L87-L93

Also, this and other code I have found seems to indicate that there should be a `boostrap` `libp2p` node that runs on the same host as the `signaling-server` - is that correct?

I was able to do peer discovery _in browser_ last time I worked with `libp2p`. Is there some breaking change that makes peer A being able to discover peer B via webrtc non-working right now in node.js? 

I want each peer using my app to auto subscribe to a specific Topic and be able to chat - via Node.js. I can get this to work in older versions of `libp2p` - specifically just by running the code here:
https://github.com/libp2p/js-libp2p-examples/blob/master/chat/nodejs/08-End/index.js as client
and 
https://github.com/libp2p/js-libp2p-examples/blob/master/chat/nodejs/bootstrap/index.js as server/client

but - for whatever (probably my dumb configuration) not in 0.27.8 (in node.js)

My app's configuration is here: https://github.com/daviddahl/wut/blob/master/p2p.js#L58-L106

I do have a dedicated server and have tried running the plain vanilla webrtc server script and the js-libp2p-examples `bootstrap/index.js` as a server, but cannot get peers to see one another.

**Comment by daviddahl** [comment]
Update: I deleted my last comment - I do think things are working now. cheers,

**Comment by autonome** [comment]
Can this be closed now, or is there a bug still?

**Comment by daviddahl** [comment]
It can be closed - everything working on my end now.

**Comment by autonome** [comment]
🎉

**Comment by autonome** [comment]
@vasco-santos is there a documentation change or update from this, or is the solution to fix #576?

**Comment by vasco-santos** [comment]
We already have the [docs](https://github.com/libp2p/js-libp2p/blob/v0.27.8/doc/CONFIGURATION.md#customizing-transports). The problem here is more the not friendly and confusing config that we have. #576 work aims to bring a better way of configuring libp2p.
